### ISSUE TITLE
evaluation mode broken for detection (both python3 and python2)

### ISSUE THREAD

**Issue by pwais** [tensorflow/tpu#516]
I have a script based upon `train_and_eval ` mode in `main.py` ( https://github.com/tensorflow/tpu/blob/master/models/official/detection/main.py ) 

Two separate issues:

1) `executor.prepare_evaluation()` will cause an outright segfault if you don't have eval stuff set up properly (e.g. specify a `val_json_file`).  There should probably be a guard to protect against unexpected configuration.  The segfault is super hard to debug!!

2) Bigger issue is that evaluation doesn't work at all!  I've tried with resnet50 and resnet152 backbones.  The checkpoint loads and then stackdriver shows that compilation starts for the TPU.  After a few minutes, stackdriver shows the node stat-ing tfrecords files and then a `docker rm tensorflow` and I see this on the client side:

```
2019-09-19 00:23:28.933396: W tensorflow/core/distributed_runtime/rpc/grpc_remote_master.cc:157] RPC failed with status = "Unavailable: Socket closed" and grpc_error_string = "{"created":"@1568852608.933229989","description":"Error received from peer","file":"external/grpc/src/core/lib/surface/call.cc","file_line":1039,"grpc_message":"Socket closed","grpc_status":14}", maybe retrying the RPC
2019-09-19 00:23:28.933396: W tensorflow/core/distributed_runtime/rpc/grpc_remote_master.cc:157] RPC failed with status = "Unavailable: Socket closed" and grpc_error_string = "{"created":"@1568852608.933254324","description":"Error received from peer","file":"external/grpc/src/core/lib/surface/call.cc","file_line":1039,"grpc_message":"Socket closed","grpc_status":14}", maybe retrying the RPC
```

Since there are no unit tests in this repo, is there any proof that this code runs for people inside of Google?  BTW I tried evaluation using both python2 and python3 per advice from @allenwang28 and I get the same outcome for both.

**Comment by allenwang28** [comment]
Can you try following the steps here? https://cloud.google.com/tpu/docs/tutorials/retinanet

Evaluation will work with those steps, so let's see if we can figure out any deviations that might be causing your failures.

**Comment by pwais** [comment]
Hi @allenwang28 for the segfault, it's straightforward to reproduce if one just does not configure an evaluation set.  You'll get the segfault as soon as `executor.prepare_evaluation()` is called.

As for the TPU bug, yes my setup is very similar to that tutorial.  (I'd like to note that the TPU version strays from the `tensorflow/models` version in requiring that TFRecords have a numeric `source_id` so that the TPU code can pipe stuff through MSCOCO's code for evaluation.  The type of `source_id` is a string and in `tensorflow/models` it's fine for that value to be a stirng, but the TPU code parses it to an `int`.  Running into this discrepancy is extremely frustrating and I highly recommend either documenting the issue conspicuously or putting in a mitigation, like hashing the string value).

I hope you appreciate, but I don't have the time or resources to download MSCOCO from scratch and to repro a bug for you.  As in my report, the crash appears with certainty to be entirely behind the TPU gRPC API, so perhaps a first step for your team is to improve logging of errors to the client.  Even if I were to follow your suggestion of running an MSCOCO experiment, due to the very poor logging available it's unlikely I could shed any new light onto the issue.

I would also heavily recommend unit tests for this repo.  Without unit tests, it's hard for users of this code to know what is expected to run and what is in a pre-alpha state.

**Comment by vidhishanair** [comment]
Hi,
Is there any progress on what causes this? I have the same "Unavailable Socket Closed" error when doing an evaluation on TPU. 
I cannot move to Python2 since I have some dependencies that work only on Python3.

**Comment by pwais** [comment]
@vidhishanair FWIW python2 did not fix anything for me.  Both python2 and python3 would crash.

**Comment by allenwang28** [comment]
For repro purposes, what script(s) were used to create the COCO dataset if not from https://github.com/tensorflow/tpu/tree/master/tools/datasets?
Was it from here?
https://github.com/tensorflow/models/blob/master/research/object_detection/dataset_tools/create_coco_tf_record.py

**Comment by pwais** [comment]
Yes, I followed the tutorial you have on the website as best I could.  I
had data that was smaller than full COCO for testing.  If you used scripts
other than those in tensorflow/models , could you please include them in
the TPU repo?  There are subtle differences in the TPU code, like image id
must be a string-ified integer.  It’s most helpful for the code in the TPU
repo to have reproducible results.  Unit tests too would be a big benefit
because then users can easily verify if something in master is broken.

**Comment by pwais** [comment]
A mini-COCO demo would also be very helpful! Perhaps your Product Manager
can help you understand, but having to pay (especially without credits) for
a COCO-sized demo is rather unattractive to most real users.

**Comment by allenwang28** [comment]
Hmm, considering that it's training correctly, I suspect that it's most likely not an issue with the subtle differences between TFRecords.

How are you passing in the eval data? As `eval_file_pattern` or `val_json_file`?

**Comment by pwais** [comment]
I tried both and also tried instrumenting the code to see what was amiss.
I strongly suggest you provide unit tests to both validate correctness as
well as help document supported usage.

It would also be very helpful if the TPU RPC errors were more decipherable
and didn’t take as much as 5-10 minutes to appear.  The main impediment to
debugging this issue has been that the failure occurs after TPU execution
starts and it takes a very very long time for the error to appear.

**Comment by pwais** [comment]
Upon trying this again, I saw this error in stackdriver:
```
W1112 03:18:29.231704 3663 op_kernel.cc:1651] OP_REQUIRES failed at transfer_ops.cc:49 : Invalid argument: Infeed shape mismatch for tuple element 65; expected [4,200], got [4,10000]
```

the error of course was not forwarded to the client :(  on the client I simply see:
```
 RPC failed with status = "Unavailable: Socket closed" and grpc_error_string = "{"created":"@1573528852.577342082","description":"Error received from peer","file":"external/grpc/src/core/lib/surface/call.cc","file_line":1039,"grpc_message":"Socket closed","grpc_status":14}", maybe retrying the RPC
```

Not sure if this is related to the previous problem, but @allenwang28 a prime example of where the TPU gRPC client is hiding very very important errors

**Comment by StKlaas** [comment]
I have the same issue as you during prediction using TPU v3. I checked the  official website and it  reveals 

> Prediction on TPU is an experimental feature to support large batch inference. It is not designed for latency-critical system. In addition, due to some usability issues, for prediction with small dataset, CPU .predict, i.e., creating a new TPUEstimator instance with use_tpu=False, might be more convenient.

This confirms that there should be some usability issues. 
My code works well in colab TPU, so I switch TPU v3 to TPU v2. Surprisely, everything works well after that.

**Comment by wmmxk** [comment]
I run into the same issue when training a BERT model using TPU. The error I run into is:  W tensorflow/core/distributed_runtime/rpc/grpc_remote_master.cc:157] RPC failed with status = "Unavailable: Socket closed" and grpc_error_string = "{"created":"@1579064280.720957631","description":"Error received from peer","file":"external/grpc/src/core/lib/surface/call.cc","file_line":1039,"grpc_message":"Socket closed","grpc_status":14}", maybe retrying the RPC

**Comment by sailordiary** [comment]
Getting GRPC errors on custom code as well. What's going on?

**Comment by Selimonder** [comment]
I am experiencing the same issue with v2-8 with TF 2.2. Does this have to do anything with TPU Quotas over TPU overload?

**Comment by bilalsattar** [comment]
any update on this issue ?
I am having same error for tensorflow/lingvo library while training their built-in models?
i am using tf 2.1 in tpu.

**Comment by mrm8488** [comment]
> 
> 
> I run into the same issue when training a BERT model using TPU. The error I run into is: W tensorflow/core/distributed_runtime/rpc/grpc_remote_master.cc:157] RPC failed with status = "Unavailable: Socket closed" and grpc_error_string = "{"created":"@1579064280.720957631","description":"Error received from peer","file":"external/grpc/src/core/lib/surface/call.cc","file_line":1039,"grpc_message":"Socket closed","grpc_status":14}", maybe retrying the RPC

Any news? I am getting the same error

**Comment by JaeminK** [comment]
Should do server.join() for all non-chief servers
